src/components/AppController.tsx

```diff
@@ -78,40 +78,73 @@ interface AppControllerProps {
-
     // States
-    const setLogEventNum = useContextStore((state) => state.setLogEventNum);
-    const logFileManagerProxy = useLogFileManagerStore((state) => state.logFileManagerProxy);
+    const beginLineNumToLogEventNum = useViewStore((state) => state.beginLineNumToLogEventNum);
+
+    const fileSrc = useLogFileStore((state) => state.fileSrc);
     const loadFile = useLogFileStore((state) => state.loadFile);
+    const setFileSrc = useLogFileStore((state) => state.setFileSrc);
+
+    const isPrettified = useViewStore((state) => state.isPrettified);
+    const updateIsPrettified = useViewStore((state) => state.updateIsPrettified);
+
+    const {logFileManagerProxy} = useLogFileManagerStore.getState();
     const numEvents = useLogFileStore((state) => state.numEvents);
-    const beginLineNumToLogEventNum = useViewStore((state) => state.beginLineNumToLogEventNum);
-    const setIsPrettified = useViewStore((state) => state.updateIsPrettified);
+
+    const logEventNum = useViewStore((state) => state.logEventNum);
+    const setLogEventNum = useViewStore((state) => state.setLogEventNum);
+
     const updatePageData = useViewStore((state) => state.updatePageData);
+
     const setUiState = useUiStore((state) => state.setUiState);
 
     // Refs
-    const isPrettifiedRef = useRef<boolean>(isPrettified ?? false);
+    const isPrettifiedRef = useRef<boolean>(isPrettified);
     const logEventNumRef = useRef(logEventNum);
 
-    // Synchronize `logEventNumRef` with `logEventNum`.
     useEffect(() => {
-        if (null !== logEventNum) {
-            logEventNumRef.current = logEventNum;
-            setLogEventNum(logEventNum);
-        }
-    }, [
-        logEventNum,
+        const handleHashChange = () => {
+            const hashParams = getWindowUrlHashParams();
+
+            if (null !== hashParams.logEventNum) {
+                setLogEventNum(hashParams.logEventNum);
+            }
+
+            if (null !== hashParams.isPrettified) {
+                updateIsPrettified(hashParams.isPrettified);
+            }
+
+            // Also check search params to handle initial page load and maintain full URL state
+            const searchParams = getWindowUrlSearchParams();
+
+            if (null !== searchParams.filePath) {
+                setFileSrc(searchParams.filePath);
+            }
+        };
+
+        handleHashChange();
+
+        window.addEventListener("hashchange", handleHashChange);
+
+        return () => {
+            window.removeEventListener("hashchange", handleHashChange);
+        };
+    }, [setFileSrc,
         setLogEventNum,
-    ]);
+        updateIsPrettified]);
 
     // Synchronize `isPrettifiedRef` with `isPrettified`.
     useEffect(() => {
-        isPrettifiedRef.current = isPrettified ?? false;
-        setIsPrettified(isPrettifiedRef.current);
+        isPrettifiedRef.current = isPrettified;
     }, [
         isPrettified,
-        setIsPrettified,
+    ]);
```

Comment:
see the comment at https://github.com/y-scope/yscope-log-viewer/pull/286/files#r2108340456 , i believe this useEffect hook can be removed